src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CartRuleController.php

```diff
@@ -261,6 +286,60 @@ public function bulkDisableAction(Request $request): RedirectResponse
         return $this->redirectToRoute('admin_cart_rules_index');
     }
 
+    /**
+     * This just prefills form fields that depends on query parameter like customer search input (if id is provided),
+     * all remaining data should be set in related form data provider.
+     * (The parameter is in the URL, not the post data which is why it's out of the provider's responsibility)
+     *
+     * @param Request $request
+     *
+     * @return array
+     */
+    private function prefillFormDataForCreation(Request $request): array
+    {
+        $formData = [];
+
+        $customerId = $request->query->getInt('customerId');
+        if ($customerId) {
+            $cartRuleFormDataProvider = $this->get(CartRuleFormDataProvider::class);
```

Comment:
This shouldn't be needed since the `FormBuilder` already calls the provider, and then merge it with the provided data. That's the whole purpose of the `$data` parameter and the `array_merge` performed here https://github.com/PrestaShop/PrestaShop/blob/5d7fd2ef9233e8a0a331e723dfa02e2574836521/src/Core/Form/IdentifiableObject/Builder/FormBuilder.php#L105

But do you mean it doesn't work because `array_merge` can't handle multidimensional arrays? If so I think it's a bug that we didn't notice because very few form controllers use this `$data` parameter and even fewer had multidimensional input (probably even none) So I think it's fine to modify/improve this in the FormBuilder

Reply:
yep it doesn't handle it well. To handle multidimensional we would need to use array_merge_recursive, but i tried it already and it breaks couple other places that way.